src/lava/lib/optimization/solvers/generic/models.py

```diff
@@ -0,0 +1,318 @@
+        # requirements
+        setattr(SolverModel, "required_resources", super_res + [CPU])
+        setattr(SolverModel, "implements_protocol", LoihiProtocol)
+        return SolverModel
+
+
+@implements(ReadGate, protocol=LoihiProtocol)
+@requires(CPU)
+class ReadGatePyModel(PyLoihiProcessModel):
+    solved: np.ndarray = LavaPyType(np.ndarray, np.int32, 32)
+    new_solution: PyInPort = LavaPyType(PyInPort.VEC_DENSE, np.int32,
+                                        precision=32)
+    do_readout: PyOutPort = LavaPyType(
+        PyOutPort.VEC_DENSE, np.int32, precision=32
+    )
+    target_cost: int = LavaPyType(int, np.int32, 32)
+    min_cost: int = None
+
+    def post_guard(self):
+        return True
+
+    def run_spk(self):
+        data = self.new_solution.recv()
+        self.do_readout.send(data)
+        self.solved[:] = data[0]
+        if data[0]:
+            print("Found a solution with cost:", self.solved)
+            self.min_cost = data[0]
+
+    def run_post_mgmt(self):
+        if self.min_cost:
+            if self.target_cost:
+                if self.min_cost <= self.target_cost:
+                    print("Best cost found:", self.min_cost)
+                    self._req_pause = True
+
+
+@implements(SolutionReadout, protocol=LoihiProtocol)
+@requires(CPU)
+class SolutionReadoutPyModel(PyLoihiProcessModel):
+    solution: np.ndarray = LavaPyType(np.ndarray, np.int32, 32)
+    read_solution: PyInPort = LavaPyType(PyInPort.VEC_DENSE, np.int32,
+                                         precision=32)
+    ref_port: PyRefPort = LavaPyType(PyRefPort.VEC_DENSE, np.int32,
+                                     precision=32)
+    target_cost: int = LavaPyType(int, np.int32, 32)
+    min_cost: int = None
+
+    def post_guard(self):
+        return True
+
+    def run_spk(self):
+        self.min_cost = None
+        data = self.read_solution.recv()
+        if data[0]:
+            self.min_cost = data[0]
+
+    def run_post_mgmt(self):
+        if self.min_cost:
+            solution = self.ref_port.read()
+            self.solution[:] = solution
+            print(f"Read solution: {self.solution}")
+            # if self.target_cost:
+            #     if self.min_cost <= self.target_cost:
+            #         self._req_pause = True
+
+
+@implements(proc=DiscreteVariablesProcess, protocol=LoihiProtocol)
+@requires(CPU)
+class DiscreteVariablesModel(AbstractSubProcessModel):
+    def __init__(self, proc):
+        # Instantiate child processes
+        # The input shape is a 2D vector (shape of the weight matrix).
+        wta_weight = -2
+        shape = proc.proc_params.get("shape", (1,))
+        diagonal = proc.proc_params.get("cost_diagonal")
+        weights = proc.proc_params.get(
+            "weights",
+            wta_weight
+            * np.logical_not(np.eye(shape[1] if len(shape) == 2 else 0)),
+        )
+        noise_amplitude = proc.proc_params.get("noise_amplitude", 1)

```

Comment:
I would strongly advice against setting default values in the PyProcModel. Rather, set them in the Process and ensure that these hyperparameters will always be defined.